feat(quic): Add draft-29 support

elenaf9 opened the pull request:
## Description

Add support for QUIC draft-29 / the `quic` codepoint.  This enables both dialing and listening on `quic` addresses.
The motivation for adding support is to allow users to connect to old go-libp2p nodes that don't support the `quic-v1` codepoint.
**Per default support is disabled** (cc @marten-seemann).

## Links to any relevant issues

See discussion in #3133.

## Open Questions

We use the same codepoint when reporting our listening addresses as address had that the user called `listen_on` on. 
If users want to report listening addresses for both codepoints they either have to manually duplicate and map each listening address, or create two listeners. 
The alternative would be to directly report listening addresses for both codepoints if draft-29 is enabled. I don't have a strong opinion, but I am leaning towards the current implementation to avoid reporting draft-29 addresses unless it is explicitly wanted by the user.

## Change checklist


- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

elenaf9 committed:
feat(quic): Add draft-29 support

thomaseizinger reviewed (commented):
Not too invasive so this looks okay to add. We may want to document that this will be removed in the future.

elenaf9 committed:
docs(quic): add note on draft-29 support removal

elenaf9 committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into quic/draft-29

mxinden commented on transports/quic/src/endpoint.rs:
```diff
@@ -280,6 +295,8 @@ pub enum ToEndpoint {
         /// UDP address to connect to.
         addr: SocketAddr,
+        /// Whether to dial the remote on QUIC version draft-29.
+        is_draft_29: bool,
```
Instead of using a `bool`, I would prefer a new enum, e.g.:

``` rust
enum QuicVersion {
  V1,
  Draft29,
}
```

What do you think? Not a strong opinion.

elenaf9 replied:
:+1: Good idea. added in 1fc87da14029b3fc338d8a717b97854924cd90fc

mxinden commented on transports/quic/src/transport.rs:
```diff
@@ -49,12 +49,23 @@ use std::{
+/// draft-29 and only supported if [`Config::support_draft_29`] is set to `true`.
+/// Note that in that case servers support both version an all QUIC listening addresses.
+///
+/// Version draft-29 should only be used to connect to nodes from other libp2p implementations
+/// that do not support `QuicV1` yet. Support for it will be removed long-term.
```
How about linking to https://github.com/multiformats/multiaddr/issues/145 here?

elenaf9 replied:
Done in fa39fc651ff49f1a173ced0c2c7cc05a57e27242.

mxinden reviewed (approved):
This looks good to me. Two suggestions, none of them blocking.

elenaf9 committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into quic/draft-29

elenaf9 committed:
chore: bump version, add changelog

elenaf9 committed:
feat(quic): add `QuicVersion` enum

elenaf9 committed:
doc(quic): link to issue about quic codepoint

thomaseizinger commented on transports/quic/src/transport.rs:
```diff
@@ -49,12 +49,24 @@ use std::{
 
 /// Implementation of the [`Transport`] trait for QUIC.
+///
+/// Per default only QUIC Version 1 (RFC 9000) is supported. In the [`Multiaddr`] this maps to
```
```suggestion
/// By default only QUIC Version 1 (RFC 9000) is supported. In the [`Multiaddr`] this maps to
```

Not a native speaker but "per" sounds odd to me.

thomaseizinger commented on transports/quic/tests/smoke.rs:
```diff
@@ -183,6 +183,98 @@ fn concurrent_connections_and_streams_tokio() {
+        let _ = d_transport.poll_next_unpin(cx);
+        Poll::Pending
+    });
+    select! {
```
@mxinden converted me to prefer the `select` fn. Can we convert you too? :)

elenaf9 replied:
Can you point me to the discussion about the reasoning behind this? Right now I have a slight preference for the macro because of the readability.

thomaseizinger replied:
Basically the rationale is:

- Plays better with IDEs
- Much easier to understand
- Clear type-safety in how to handle the various outcomes
- No fuse needed
- No magic syntax to learn

I don't know where exactly we had this discussion, so can't link you to it unfortunately.

mxinden replied:
While I very much believe in the cause @thomaseizinger, I suggest we delay further select-evangelizing towards @elenaf9 and others to future pull request. In other words, I don't think we need to block here. (Also gives us more time to print t-shirts and stickers.)

thomaseizinger commented on transports/quic/tests/smoke.rs:
```diff
@@ -183,6 +183,98 @@ fn concurrent_connections_and_streams_tokio() {
+                p
+            }
+        })
+        .collect();
```
This snippet is repeated 3 times in the test. Can we make a little `swap_protocol` utility so these are one-liners?

elenaf9 replied:
Wdyt of 4cd494d?

thomaseizinger commented on transports/quic/src/transport.rs:
```diff
@@ -486,6 +511,12 @@ impl<P: Provider> Drop for Listener<P> {
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum QuicVersion {
```
```suggestion
pub enum ProtocolVersion {
```

elenaf9 replied:
Done in 8f05e09.

thomaseizinger commented on transports/quic/src/transport.rs:
```diff
@@ -486,6 +511,12 @@ impl<P: Provider> Drop for Listener<P> {
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum QuicVersion {
+    V1,
```
```suggestion
    V1, // i.e. RFC9000
```

thomaseizinger reviewed (approved):
LGTM

Only a few nits!

thomaseizinger commented on transports/quic/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "libp2p-quic"
-version = "0.7.0-alpha"
+version = "0.7.1-alpha"
```
I think the appropriate bump would be:


```suggestion
version = "0.7.0-alpha.2"
```

`0.7.1-alpha` suggests that a `0.7.0` exists and we are pre-releasing a patch version (although the non-breaking aspect of said patch version is not guaranteed, hence the alpha).

We haven't even released `0.7.0` so we may as well push more alpha's out until we are happy with it.

elenaf9 replied:
Ah yes you are right! Fixed in 2a83d3c. This also means that we don't need to bump the version in the `libp2p` crate.

thomaseizinger replied:
> Ah yes you are right! Fixed in 2a83d3c. This also means that we don't need to bump the version in the `libp2p` crate.

I think you will have to, I don't think alpha versions update automatically. AFAIK, they are basically major version bumps but with an "unstable" co-notation.

thomaseizinger replied:
I could be wrong though, please test it :)

elenaf9 committed:
fix(quic): fix copy-pase errors in smoke test

elenaf9 committed:
feat(quic): add swap_proto macro in smoke test

elenaf9 committed:
refactor(quic): rename QuicVersion => ProtocolVersion

elenaf9 committed:
chore(quic): relase patch as 0.7.0-alpha.2

Change version to 0.7.0-alpha.2 rather than 0.7.1-alpha. The latter
wrongly implies that we are pre-releasing 0.7.1.
Revert version bump and changelog entry in `libp2p` crate.

elenaf9 committed:
fix(quic): fix version in CHANGELOG entry

elenaf9 committed:
doc(quic): apply suggestions from review

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

elenaf9 committed:
Merge branch 'master' into quic/draft-29

mxinden reviewed (approved):
Changes and version handling looks good to me.

mxinden committed:
Merge branch 'master' into quic/draft-29
